### PR TITLE
MINIFICPP-543: Implement C2 based update policy enforcement

### PR DESCRIPTION
**Pull request by phrocker**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by apiri** [comment]
reviewing

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace controllers {
+
+core::Property UpdatePolicyControllerService::EnableAllProperties("Enable All Properties", "Enables all properties", "false");
+core::Property UpdatePolicyControllerService::AllowedProperties("Allowed Properties", "Properties for which we will allow updates");
+core::Property UpdatePolicyControllerService::DisallowedProperties("Disallowed Properties", "Properties for which we will not allow updates");
+core::Property UpdatePolicyControllerService::PersistUpdates("Persist Updates", "Property that dictates whether updates should persist a restart");
```
Comment (apiri) [inline libminifi/src/controllers/UpdatePolicyControllerService.cpp]: a -> after

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+void UpdatePolicyControllerService::onEnable() {
+  std::string enableStr, persistStr;
+
+  bool enable_all = false;
+  if (getProperty(EnableAllProperties.getName(), enableStr)) {
+    enable_all = utils::StringUtils::StringToBool(enableStr, enable_all);
+  }
+
+  if (getProperty(PersistUpdates.getName(), persistStr)) {
```
Comment (apiri) [inline libminifi/src/controllers/UpdatePolicyControllerService.cpp]: I saw this is being handled via MINIFICPP-544.  It seems like this should be used in conjunction with the temp/permanent considerations.  Scanned through the C2 wiki page but am a little hazy about how this interacts with temp/permanent.  I assume the policy service would get precedence but am uncertain of conditions such as I request allowing permanent but the instance is not configured to persist config changes.

File: libminifi/test/unit/UpdatePolicyTests.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+TEST_CASE("TestEnableAll", "[test1]") {
+  auto controller = std::make_shared<minifi::controllers::UpdatePolicyControllerService>("TestService");
+  std::shared_ptr<minifi::Configure> configuration = std::make_shared<minifi::Configure>();
+  controller->initialize();
+  controller->setProperty(minifi::controllers::UpdatePolicyControllerService::EnableAllProperties, "true");
+  controller->onEnable();
+  REQUIRE(true == controller->canUpdate("anyproperty"));
+}
+
+TEST_CASE("TestEnableAllFalls", "[test1]") {
```
Comment (apiri) [inline libminifi/test/unit/UpdatePolicyTests.cpp]: Falls -> Fails

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    persist_updates_ = utils::StringUtils::StringToBool(persistStr, persist_updates_);
+  }
+
+  auto builder = state::UpdatePolicyBuilder::newBuilder(enable_all);
+
+  core::Property all_prop("Allowed Properties", "Properties for which we will allow updates");
+  core::Property dall_prop("Disallowed Properties", "Properties for which we will allow updates");
```
Comment (apiri) [inline libminifi/src/controllers/UpdatePolicyControllerService.cpp]: allow -> disallow

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+#endif
+#include "core/state/UpdatePolicy.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace controllers {
+
+core::Property UpdatePolicyControllerService::EnableAllProperties("Enable All Properties", "Enables all properties", "false");
```
Comment (apiri) [inline libminifi/src/controllers/UpdatePolicyControllerService.cpp]: Not sure I am tracking the delineation between "enable" and "allow".  It seems that enable all might be better suited as AllowAll?

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp (hunk 1 above)
Comment (phrocker) [reply libminifi/src/controllers/UpdatePolicyControllerService.cpp]: This is the intransitive verb.

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp (hunk 2 above)
Comment (phrocker) [reply libminifi/src/controllers/UpdatePolicyControllerService.cpp]: "I assume the policy service would get precedence but am uncertain of conditions such as I request allowing permanent but the instance is not configured to persist config changes."

I'm not sure I follow...but perhaps I should explain this a little better before adding a comment for the code to ensure it makes sense. Updates can be controlled via the flow via the controller service. There may be cases where you do not want configuration updates to be persisted. An easy case would be one in which there is no volatile storage. That's a bit trite as it over exemplifies the argument; however, a less whimsical example may be one in which you only want updates to persist through transferring the minifi.properties file. This ONLY controls updates performed through mechanisms that use the controller service. Does that make sense? If so I'd like to document that on the wiki and in the code as your question points out that it's a point of confusion.

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp (hunk 4 above)
Comment (phrocker) [reply libminifi/src/controllers/UpdatePolicyControllerService.cpp]: COPY PASTA! Thanks!

File: libminifi/test/unit/UpdatePolicyTests.cpp (hunk 3 above)
Comment (phrocker) [reply libminifi/test/unit/UpdatePolicyTests.cpp]: Oh wow. Thanks. This is what I get for using a 4k monitor.

File: libminifi/src/controllers/UpdatePolicyControllerService.cpp (hunk 5 above)
Comment (phrocker) [reply libminifi/src/controllers/UpdatePolicyControllerService.cpp]: I typed a long response that Allow All doesn't fit what I intended with disallow, below; however, it was lost. I think the gist of that response was I will think about a better name to convey its meaning.

**Comment by phrocker** [comment]
@apiri I made some updates to the wiki entry.

**Comment by apiri** [comment]
@phrocker thank you for the updates, definitely clears things up.  will get this merged in

**Pull request closed without merging**